src/core/ddsc/tests/psmxif.c

```diff
@@ -0,0 +1,441 @@
+    CU_ASSERT_FATAL((props & DDS_DATA_TYPE_IS_MEMCPY_SAFE) == DDS_DATA_TYPE_IS_MEMCPY_SAFE);
+    props = dds_stream_data_types(PsmxType1_desc.m_ops);
+    CU_ASSERT_FATAL((props & DDS_DATA_TYPE_IS_MEMCPY_SAFE) == DDS_DATA_TYPE_IS_MEMCPY_SAFE);
+  }
+
+  const dds_domainid_t domainId = 0;
+  bool supports_shared_memory_expected = true;
+  dds_psmx_endpoint_t* psmx_endpt_expected = NULL;
+  dds_psmx_topic_t* psmx_topic_expected = NULL;
+  bool specify_locator[] = {true, false};
+  uint8_t locator_in[16];
+  memset(locator_in, 0x0, sizeof(locator_in)); // Avoid warning 'uninitialized value'.
+  ((uint64_t*)locator_in)[0] = (uint64_t)0x4a4d203df6996395;
+  ((uint64_t*)locator_in)[1] = (uint64_t)0xe1412fbecc2de4b6;
+
+  int N = sizeof(specify_locator) / sizeof(specify_locator[0]);
+  for (int i = 0; i < N; ++i)
+  {
+    const uint8_t* locator = specify_locator[i] ? locator_in : NULL;
+    char* configstr_in = create_config(domainId, "dummy", locator);
+    const dds_entity_t domain = dds_create_domain(domainId, configstr_in);
+    CU_ASSERT_FATAL(domain > 0);
+    const dds_entity_t participant = dds_create_participant(domainId, NULL, NULL);
+    CU_ASSERT_FATAL(participant > 0);
+    {
+      // Query whether shared memory is supported.
+      dds_domain* dom = NULL;
+      {
+        dds_entity* x = NULL;
+        dds_return_t rc = dds_entity_pin(domain, &x);
+        CU_ASSERT_FATAL(rc == DDS_RETCODE_OK && dds_entity_kind(x) == DDS_KIND_DOMAIN);
+        dom = (dds_domain*)x;
+      }
+      assert(dom->psmx_instances.length >= 1);
+      struct dummy_psmx* dpsmx = (struct dummy_psmx*)dom->psmx_instances.instances[0];
+      dmock = dpsmx->mockstats_get_ptr();
+      CU_ASSERT_FATAL(dmock->cnt_create_psmx == 1); // Confirm the dummy psmx has been loaded.
+      dds_entity_unpin(&dom->m_entity);
+    }
+    // Check that the config string passed to `dds_create_domain()` has been correctly forwarded to the dummy psmx.
+    CU_ASSERT_FATAL(strstr(configstr_in, dmock->config) != NULL); // dmock->config is a substring of the original xml.
+    ddsrt_free(configstr_in);
+
+    dds_entity_t writer1 = 0, reader1 = 0, writer2 = 0, reader2 = 0;
+    {
+      char topicname[100];
+      dummy_topics_alloc(dmock, 2);
+      dummy_endpoints_alloc(dmock, 4);
+
+      create_unique_topic_name("shared_memory", topicname, sizeof(topicname));
+      psmx_topic_expected = (dds_psmx_topic_t*)dmock->topics._buffer + dmock->topics._length;
+      dds_entity_t topic1 = dds_create_topic(participant, &SC_Model_desc, topicname, NULL, NULL);
+      CU_ASSERT_FATAL(topic1 > 0);
+      {
+        dds_entity* x = NULL;
+        dds_return_t rc = dds_entity_pin(topic1, &x);
+        CU_ASSERT_FATAL(rc == DDS_RETCODE_OK && dds_entity_kind(x) == DDS_KIND_TOPIC);
+        struct dds_psmx_topics_set* topics_set = &((dds_topic*)x)->m_ktopic->psmx_topics;
+        CU_ASSERT_FATAL(topics_set->length == 1 && topics_set->topics[0] == psmx_topic_expected);
```

Comment:
This is one way of verifying that the returned address is picked up correctly. Still, I prefer my original suggestion: we know the address must be passed to any `dummy_psmx_create_endpoint` call that is expected based on creating an endpoint of topic `topic1`.

And so rather I would much rather not dig so deeply in the internals and instead check in `dummy_psmx_create_endpoint` that the expectations are met. That eliminates a fair amount of ugly code here.